File: my-folder/0003-longest-substring-without-repeating-characters/solution.c
int lengthOfLongestSubstring(char * s){
    int hash_map[256] = {0};
    int left = 0, right = 0;
    int res = 0;
    while (s != NULL && s[right] != '\0'){
        char r = s[right];
        if (hash_map[r] > 0){
            if (hash_map[r] >= left)
                left = hash_map[r];
        }
        int new_len = right - left + 1;
        if (new_len > res)
            res = new_len;
        right++;
        hash_map[r] = right;
    }
    return res;
}

File: my-folder/0003-longest-substring-without-repeating-characters/solution.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        use std::collections::HashMap;
        let mut char_index_map = HashMap::new();
        let mut max_length = 0;
        let mut start = 0;

        for (i, c) in s.chars().enumerate() {
            if let Some(prev_index) = char_index_map.insert(c, i) {
                if prev_index >= start {
                    start = prev_index + 1;
                }
            }
            max_length = max_length.max(i - start + 1);
        }

        max_length as i32
    }
}
